### ISSUE TITLE
Helm v2.16: no matches for kind "Deployment" in version "apps/v1beta1"

### ISSUE THREAD

**Issue by discostur** [helm/helm#6969]
Hi,

i know that k8s v1.16 deprecated the old API versions - but i don't know how helm handles this. I did the following:

* installed k8s v1.15.x cluster
* install stable/fluentd with API version **apps/v1beta1**
* upgrade k8s -> 1.16.x

Now i try to update the fluentd chart / deployment which was upgraded to the new api versions already (**apps/v1** https://github.com/helm/charts/blob/master/stable/fluentd/templates/deployment.yaml) and i get the following error:

```
helm upgrade --install --recreate-pods fluentd stable/fluentd

UPGRADE FAILED
Error: failed decoding reader into objects: unable to recognize "": no matches for kind "Deployment" in version "apps/v1beta1"
Error: UPGRADE FAILED: failed decoding reader into objects: unable to recognize "": no matches for kind "Deployment" in version "apps/v1beta1"
```

I can now delete the helm fluentd chart and reinstall it but that seems not to be the correct way ... is there any migration path?

Output of `helm version`:
```
Client: &version.Version{SemVer:"v2.16.0", GitCommit:"e13bc94621d4ef666270cfbe734aaabf342a49bb", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.16.0", GitCommit:"e13bc94621d4ef666270cfbe734aaabf342a49bb", GitTreeState:"clean"}
``` 

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:09:08Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by bacongobbler** [comment]
can you please test with 2.16.1? This was addressed in that release. Thanks.

https://github.com/helm/helm/issues/6894

**Comment by discostur** [comment]
@bacongobbler i tried with helm v2.16.1 - still the same error

**Comment by bacongobbler** [comment]
Oh! This is because you're deploying on Kubernetes 1.16, which dropped all workload APIs other then apps/v1, including apps/v1beta1.

https://kubernetes.io/blog/2019/09/18/kubernetes-1-16-release-announcement

**Comment by discostur** [comment]
Yes i know - but when you check the Chart it is already on apps/v1:

https://github.com/helm/charts/blob/master/stable/fluentd/templates/deployment.yaml

**Comment by bacongobbler** [comment]
When an upgrade occurs, the previous release (source) is compared to the new release (target). Manifests are validated against the Kubernetes API, including old versions, which is why you're geting caught up in the issue.

To fix this, downgrade to an older version of kubernetes, and you'll be able to work with v1beta1 objects again.

It's important to know that objects cannot be upgraded in-place from one apiVersion to the next - that is considered a breaking change. I would file a ticket upstream with the fluentd folks to note that changing from v1beta1 to v1 was a breaking change and cannot be upgraded. See also https://github.com/helm/helm/issues/6583.

Hope this helps.

**Comment by discostur** [comment]
ok thanks for the explanation - so i need to delete and re-install because i dondt want to downgrade k8s cluster

**Comment by yoshz** [comment]
Another solution is to add the following argument to `/etc/kubernetes/manifest/kube-apiserver.yaml` when you are running Kubernetes 1.16 or 1.17:
```
--runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true
```

**Comment by travisghansen** [comment]
For anyone else stumbling upon this, we hit a similar situation where charts were not updated *before* a new version of k8s hit. We also did not want to delete the release (ie: actually remove running/functioning resources). Our solution was to:

1. Remove all old configmaps for old revisions (effectively removing all history it ever existed from helm's perspective)
1. Run `helm install --replace --no-hooks ...` once to get the release to a `FAILED` state (it fails because it sees resources that already exist)
1. Run step 2 a second time to get a good deployment
1. Remove all configmaps again (ie: revision 1) except the last good revision
1. Run `helm upgrade --install ...` as usual

Alternatively, if you want to retain the same sequencing on the revision numbers do *something* to get a revision to fail, then start at step 3.

**Comment by rmetzler** [comment]
@travisghansen thank you very much! In my case the error message was:

```
Error: UPGRADE FAILED: failed decoding reader into objects: unable to recognize "": no matches for kind "Certificate" in version "certmanager.k8s.io/v1alpha1"
```

I deployed with `helm install --replace --name $RELEASE` and then removed all older configmaps from the `kube-system` namespace. The next deployment with `helm upgrade --install` worked for me.

**Comment by albanm** [comment]
@travisghansen  I tried doing this, but `helm install --replace` gives me `Error: cannot re-use a name that is still in use`. Someone has an idea to solve that ?

This issue is kind of a nightmare honestly. Big chunk of cluster is stuck: impossible to change any chart config, impossible to downgrade kubernetes with my provider, impossible to change kubernetes-apiserver.yaml, I don't want to risk deleting everything then reattach pvcs and stuff. This kubernetes update of a "minor" version is a mess :( Sorry for the rant !

**Comment by travisghansen** [comment]
@albanm yeah, it's a rough situation to hit for sure :( I'm not sure why you'd get that error when using `--replace`. Can you share what `helm ls/history` show for the problem release?

**Comment by albanm** [comment]
Sure,

I've got multiple releases with this issue, one of them based on incubator/kafka for example.

```
$ helm ls --filter kafka
NAME 	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART       	APP VERSION
kafka	default  	1       	2019-09-18 19:34:58.571948322 +0000 UTC	deployed	kafka-0.17.1	5.0.1      
```

```
$ helm history kafka
REVISION	UPDATED                 	STATUS  	CHART       	APP VERSION	DESCRIPTION     
1       	Wed Sep 18 19:34:58 2019	deployed	kafka-0.17.1	5.0.1      	Install complete

```

```
$ helm upgrade --install -f helmconfigs/env/kafka.yaml kafka incubator/kafka
Error: UPGRADE FAILED: unable to build kubernetes objects from current release manifest: unable to recognize "": no matches for kind "StatefulSet" in version "apps/v1beta1"
```

```
$ helm install --replace --no-hooks -f helmconfigs/env/kafka.yaml kafka incubator/kafka
Error: cannot re-use a name that is still in use
```

Kubernetes is version 1.16.4 and helm version 3.0.2.

I just realized I mixed up your steps, the next comment mentions cleaning old config maps after running `helm install --replace` but you said it should be done before. How do you do that ?

**Comment by travisghansen** [comment]
@albanm ah, I'm still using helm 2.x so steps may vary slightly (and that could be why you're getting the error even when using `--replace`). The releases are just stored as configmaps (or secrets). So just list the configmaps in your ns and it should be pretty obvious which ones are applicable.

I'd send over the script I use to clean them up but it's build for helm 2.x and so it's likely not much help for your scenario. Maybe someone else on the thread is using helm 3 and can comment on that :(

**Comment by albanm** [comment]
Ah, I looked at configmaps before and didn't see anything, its all in secrets.

Well things are not sorted yet... I hit https://github.com/helm/helm/issues/6850

But enough chat on this issue. Thanks for your help !

**Comment by itkroplis** [comment]
> Another solution is to add the following argument to `/etc/kubernetes/manifest/kube-apiserver.yaml` when you are running Kubernetes 1.16 or 1.17:
> 
> ```
> --runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true
> ```

What do you do after kube-apiserver.yaml edit? How do you apply changes?
kubeadm init  !? Or another way?

**Comment by yoshz** [comment]
> > Another solution is to add the following argument to `/etc/kubernetes/manifest/kube-apiserver.yaml` when you are running Kubernetes 1.16 or 1.17:
> > ```
> > --runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true
> > ```
> 
> What do you do after kube-apiserver.yaml edit? How do you apply changes?
> kubeadm init !? Or another way?

Changes in the manifests are automatically detected by Kubelet which will stop the related container and start a new one.

**Comment by nirupbbnk** [comment]
> Another solution is to add the following argument to `/etc/kubernetes/manifest/kube-apiserver.yaml` when you are running Kubernetes 1.16 or 1.17:
> 
> ```
> --runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true
> ```

this worked for me

**Comment by xkdidl** [comment]
> Another solution is to add the following argument to `/etc/kubernetes/manifest/kube-apiserver.yaml` when you are running Kubernetes 1.16 or 1.17:
> 
> ```
> --runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true
> ```

this change is good for me too. :)

**Comment by niranjan94** [comment]
I was able to get around this by directly editing the helm release info file stored as a kubernetes secret. 

1. We get the release json by base64 decoding twice and decompressing the bytes with gzip.

```bash
kubectl get secrets sh.helm.release.v1.shared.v4 -n shared -o jsonpath='{.data.release}' | base64 -D | base64 -D | gzip -d > release.json
```

2. `release.json` can be edited now to update all references to `apps/v1beta1`
3. Re-encode the release.json file and update the base64 value in the secret

```bash
cat release.json | gzip | base64 | base64
```

Now we're able to update the helm app as usual.

**Comment by hickeyma** [comment]
@niranjan94 Doc will be merged soon for Helm v2: #8135. There is also the [mapkubeapis plugin](https://github.com/hickeyma/helm-mapkubeapis) which automates the steps for v2 and v3.

Current open issue with most details is #7219

**Comment by niranjan94** [comment]
@hickeyma thanks for the info ! Didn't know there was an official doc for this 😄

**Comment by f-ld** [comment]
There's also now [a page for this topic in the official helm doc](https://helm.sh/docs/topics/kubernetes_apis/)

**Comment by hickeyma** [comment]
FYI, the Helm v2 doc is [here](https://github.com/helm/helm/blob/dev-v2/docs/kubernetes_apis.md). It will be rendered into the [v2 online docs](https://v2.helm.sh/docs/) when v2.16.8 is released.

**Comment by itkroplis** [comment]
> > > Another solution is to add the following argument to `/etc/kubernetes/manifest/kube-apiserver.yaml` when you are running Kubernetes 1.16 or 1.17:
> > > ```
> > > --runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true
> > > ```
> > 
> > 
> > What do you do after kube-apiserver.yaml edit? How do you apply changes?
> > kubeadm init !? Or another way?
> 
> Changes in the manifests are automatically detected by Kubelet which will stop the related container and start a new one.

Hi! I update my K8s to 1.18.2 but this line - --runtime-config=apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,...    Unfortunately it doesn't work anymore. :(

**Comment by WoodProgrammer** [comment]
Hi,

nowadays I'm working in a similar case. My application was working on 1.15 based cluster with helm version 2.11. Before upgrade the cluster version to 1.16 I have updated apiVersion of objects like RBAC, Deployment, Ingress .. etc and updated RELEASE with helm upgrade easily.

To verify it you can check with `helm get RELEASE_NAME `command and you should see the updated version of deployment as `apps/v1`.

After completed this progress, upgraded cluster version to 1.16 and currently I'm working deployment progress.

I hope this may help you. before testing it you should run the same flow on your sandbox environment.

**Comment by fl-max** [comment]
Can confirm; the [official Helm docs](https://helm.sh/docs/topics/kubernetes_apis/#updating-api-versions-of-a-release-manifest) and the [mapkubeapis plugin](https://github.com/hickeyma/helm-mapkubeapis) resolved this for me.

**Comment by juan-vg** [comment]
Kubernetes: 1.16
Helm&Tiller: 2.16.1

Upgrade was failing due to `Error: unable to recognize "": no matches for kind "StatefulSet" in version "apps/v1beta1"
` on a **PROD** cluster. Reinstalling was not an option.

[The mapkubeapis plugin](https://github.com/hickeyma/helm-mapkubeapis) definitely is a simple and working solution. Works like a charm. My most sincere respects to its developers!! :sunglasses: 

PD: I got the helm error on a plan step (dry-run), so I hadn't to remove any failing release. Just install plugin and run it
```bash
helm plugin install https://github.com/hickeyma/helm-mapkubeapis
helm mapkubeapis --v2 my-release-name
```
And then the upgrade run without any issue!

**Comment by sescotti** [comment]
@travisghansen amazing, it worked like a charm!

In case someone has a release with a high revision number, deleting them one by one can be a PITA, so here's a one-liner to delete them all -- just replace "release-name" with your...release name :) 

`kubectl get configmap -n kube-system -o json | jq '.items[].metadata.name | select(contains("release-name")) | "kubectl delete configmap -n kube-system \(.)"' | xargs -n1 bash -c`

**Comment by jsokol89** [comment]
I bumped into the same issue.
My solution was to patch helm's ConfigMap for the affected release.

**TL;DR:**
To do so, you have to take ".data.release" of the ConfigMap causing issues, base64 decode it, un-gzip and the de-protobuf it. Once done, you will see raw helm charts attached to the release. Once modified, you just roll the protobuf -> gzip -> base64 train back and you're ready to patch helm's ConfigMap. Always take backups of things you modify and remember that you do it on your own risk. For the record - I've done it heaps of times and it never caused major dramas.

**Step by step process:**

1. You need to have _protoc_ installed (https://google.github.io/proto-lens/installing-protoc.html) as well as clone the protobuf definitions as per below:
`git clone https://github.com/kubernetes/helm ~/src/helm`
`git clone https://github.com/google/protobuf ~/src/protobuf`

2. Get the raw data from helm release that requires a patch (replace ${RELEASE_NAME} and ${VERSION} to suit your needs)

`kubectl get cm -n kube-system -o jsonpath="{.data.release}" ${RELEASE_NAME}.${VERSION} | base64 -D | zcat | protoc -I ~/src/helm/_proto/ -I ~/src/protobuf/src/ --decode hapi.release.Release ~/src/helm/_proto/hapi/**/* > /tmp/release`

3. Modify _/tmp/release_ file to replace all the beta api versions for affected deployments with `apps/v1`.

4. Once done, wrap the data back to helm-friendly state with:

`cat /tmp/release | protoc -I ~/src/helm/_proto/ -I ~/src/protobuf/src/ --encode hapi.release.Release ~/src/helm/_proto/hapi/**/* | gzip | base64`

5. Copy the generated base64 string and replace the .data.release for the helm ConfigMap with either:
`kubectl edit cm -n kube-system ${RELEASE_NAME}.${VERSION}`
or equivalent `kubectl patch` that I never remember the syntax for :)

Hope it helps.

**Comment by oldthreefeng** [comment]
> Kubernetes: 1.16
> Helm&Tiller: 2.16.1
> 
> Upgrade was failing due to `Error: unable to recognize "": no matches for kind "StatefulSet" in version "apps/v1beta1" ` on a **PROD** cluster. Reinstalling was not an option.
> 
> [The mapkubeapis plugin](https://github.com/hickeyma/helm-mapkubeapis) definitely is a simple and working solution. Works like a charm. My most sincere respects to its developers!! 😎
> 
> PD: I got the helm error on a plan step (dry-run), so I hadn't to remove any failing release. Just install plugin and run it
> 
> ```shell
> helm plugin install https://github.com/hickeyma/helm-mapkubeapis
> helm mapkubeapis --v2 my-release-name
> ```
> 
> And then the upgrade run without any issue!

greate plugin !! 

```
$ helm plugin install https://github.com/hickeyma/helm-mapkubeapis
# namespace default is kube-system. tiiler pod exist namespace 
$ helm mapkubeapis --v2 my-release-name  --namespace tiller-pod-namespace
```

**Comment by hickeyma** [comment]
FYI, this is documented (including the manual steps/process) in:

- v3: https://helm.sh/docs/topics/kubernetes_apis/
- v2: https://v2.helm.sh/docs/using_helm/#deprecated-kubernetes-apis
